system-addon/test/functional/mochitest/browser_highlights_section.js

```diff
@@ -31,15 +31,11 @@ add_task(async function getHighlights() {
   await add_highlights_bookmark(count);
 
   let selector = "[data-mochitest='highlights-card']";
-  let selectorPlaceholder = "[data-mochitest='highlights-card-placeholder']";
   let message = "found cards";
 
   // it should display two highlights cards instead of placeholders.
   await check_highlights_elements(selector, count, message);
 
-  // it should display a placeholder card.
-  await check_highlights_elements(selectorPlaceholder, 1, "found placeholder");
```

Comment:
Why is this failing in the first place? An initial glance would make it seem quite odd why one assert would pass and not the other because it would seem to imply there's 2 highlights and 3 placeholders. But looking at `check_highlights_elements` shows that it's effectively doing a completely new test with opening a second new tab. Is that intended? If we checked that the page has 2 highlights and 1 placeholder, would it pass?

But even if a page is actually consistent in that there's 3 cards shown, it's quite suspicious that the `head` test functions results in this inconsistency. Is it necessary that the pref needs to be toggled each time a test runs?
```js
Services.prefs.setBoolPref("browser.newtabpage.activity-stream.feeds.section.highlights", false);
Services.prefs.setBoolPref("browser.newtabpage.activity-stream.feeds.section.highlights", true);
```

It's also a bit concerning that there's 3 very similar `head` functions that seem to have copy/pasted some open/wait/check/cleanup that each need to be fixed up if there is an underlying problem leading to this inconsistency.

Reply:
> If we checked that the page has 2 highlights and 1 placeholder, would it pass?
Haven't tested, locally you can see it showing up correctly.
The idea was to have functions that did simple assertions that work independently (like commenting out the other code in the test case).
>  Is it necessary that the pref needs to be toggled each time a check runs?
No. Only once, after history/bookmarks have been updated. But the goal was to have the assert function work regardless of any other statements/changes. This is the workaround for getting the feeds to update.
> there is an underlying problem leading to this inconsistency.
I ran all the tests with `--verify` which is supposed to specifically find these problems
```
::: 1. Run each test 10 times in one browser. : Pass
::: 2. Run each test 5 times in a new browser each time. : Pass
::: 3. Run each test 10 times in one browser, in chaos mode. : Pass
::: 4. Run each test 5 times in a new browser each time, in chaos mode. : Pass 
```
It helped me find issues regarding cleanup in these tests.
I'm not saying the code is great, just that the choices were intentional. I'm currently getting a debug build to test this.

Reply:
I was able to reproduce this locally on a debug build with `--disable-e10s`. I was expecting `--verify` included this.